homeassistant/components/automation/template.py

```diff
@@ -0,0 +1,70 @@
+"""
+homeassistant.components.automation.template
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Offers template automation rules.
+
+For more details about this automation rule, please refer to the documentation
+at https://home-assistant.io/components/automation/#template-trigger
+"""
+import logging
+
+from homeassistant.const import CONF_VALUE_TEMPLATE
+from homeassistant.exceptions import TemplateError
+from homeassistant.helpers.event import track_state_change
+from homeassistant.util import template
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def trigger(hass, config, action):
+    """ Listen for state changes based on `config`. """
+    value_template = config.get(CONF_VALUE_TEMPLATE)
+
+    if value_template is None:
+        _LOGGER.error("Missing configuration key %s", CONF_VALUE_TEMPLATE)
+        return False
+
+    # Get all entity ids
+    all_entity_ids = hass.states.entity_ids()
```

Comment:
Entity ids can change over time. By listening to the entity ids that are available on boot, we will miss state changes of future entities (ie ones that are discovered later). A better approach is to listen to [`EVENT_STATE_CHANGED`](https://github.com/balloob/home-assistant/blob/dev/homeassistant/const.py#L33).

```
def event_listener(event):
    […]

hass.bus.listen(EVENT_STATE_CHANGED, event_listener)
```